### PR TITLE
8255387: Japanese characters were printed upside down on AIX

### PR DESCRIPTION
**Pull request by toshiona**
Hi,

Could you review this fix? Under some special conditions, non-English characters were printed upside down.

At printing with raster image, the image was generated from bottom to top. So, each characters should also be drawn as vertical mirrored. However, freetype doesn't support to transform it if the font is using embedded bitmap and non-English. As the result, these Japanese characters were printed as upside down.

In this case, freetype should be prevented to use embedded bitmap.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8255387](https://bugs.openjdk.java.net/browse/JDK-8255387): Japanese characters were printed upside down on AIX

### Reviewers
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to 5bd7d84bbd0d57503e728f59ba9e6cae2dbc554b
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1218/head:pull/1218`
`$ git checkout pull/1218`

### CODE REVIEW

File: test/jdk/java/awt/print/PrinterJob/PrintTranslateTest.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import java.awt.Color;
+import java.awt.Font;
+import java.awt.FontMetrics;
+import java.awt.Graphics2D;
+import java.awt.GraphicsEnvironment;
+import java.awt.RenderingHints;
+import java.awt.geom.AffineTransform;
+import java.awt.image.BufferedImage;
+
+public class PrintTranslateTest{
```
Comment (mrserb) [inline test/jdk/java/awt/print/PrinterJob/PrintTranslateTest.java:40]: Does it really bind to the printing/"print/PrinterJob", look like the bug is reproduced in the common rendering as well?

Comment (toshiona) [reply test/jdk/java/awt/print/PrinterJob/PrintTranslateTest.java]: Thank you Sergey for your advice. I moved the test case under test/jdk/java/awt/font and changed the name. The original problem is at printing, but yes it's common rendering.

**Comment by victordyakov** [comment]
@mrserb @prrace please review

**Review by prrace — CHANGES_REQUESTED** [review]

File: test/jdk/java/awt/print/PrinterJob/PrintTranslateTest.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test PrintTranslateTest
+ * @bug 8255387
+ * @summary Vertial mirrored characters should be drawn correctly
```
[the remark this answers is not in the document]
Comment (toshiona) [reply test/jdk/java/awt/print/PrinterJob/PrintTranslateTest.java]: sorry, changed the summary description.

File: test/jdk/java/awt/print/PrinterJob/PrintTranslateTest.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import java.awt.Font;
+import java.awt.FontMetrics;
+import java.awt.Graphics2D;
+import java.awt.GraphicsEnvironment;
+import java.awt.RenderingHints;
+import java.awt.geom.AffineTransform;
+import java.awt.image.BufferedImage;
+
+public class PrintTranslateTest{
+    static String target = "\u3042";
```
[the remark this answers is not in the document]
Comment (toshiona) [reply test/jdk/java/awt/print/PrinterJob/PrintTranslateTest.java]: fixed

File: test/jdk/java/awt/print/PrinterJob/PrintTranslateTest.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    static final int SIZE = 50;
+    static final int LIMIT = 40;
+
+    static BufferedImage drawNormal(Font font) {
+        BufferedImage image = new BufferedImage(SIZE, SIZE,
+                                      BufferedImage.TYPE_BYTE_BINARY);
```
[the remark this answers is not in the document]
Comment (toshiona) [reply test/jdk/java/awt/print/PrinterJob/PrintTranslateTest.java]: yes

File: src/java.desktop/share/native/libfontmanager/freetypeScaler.c [hunk 5]
```diff
@@ -529,7 +529,8 @@ Java_sun_font_FreetypeFontScaler_createScalerContextNative(
      */
     if ((aa != TEXT_AA_ON) && (fm != TEXT_FM_ON) &&
         !context->doBold && !context->doItalize &&
-        (context->transform.yx == 0) && (context->transform.xy == 0))
+        (context->transform.yx == 0) && (context->transform.xy == 0) &&
+        (context->transform.yy > 0))
```
[the remark this answers is not in the document]
Comment (toshiona) [reply src/java.desktop/share/native/libfontmanager/freetypeScaler.c]: Thank you for the advice. Updated the patch and test to cover horizontal mirror case.

**Review by prrace — APPROVED** [review]

**Comment by prrace** [comment]
Approved.
Whilst the fix looks reasonable and important for on-screen, I am surprised it was an issue when printing.
At *printer resolution* it is really surprising that there are any bitmaps for the glyphs.
We should be talking about a 300 dpi printer minimum these days and so we should be requesting at least 50 pixel
high text for even 12pt nomimal size. Fonts usually don't have bitmaps at that size because they are large  and less important.
So I am not sure under what circumstances you saw it.

Secondly, I think that unless we were going down the raster path (which I think you said we were) that we normally explicitly used filled shapes for CJK.

**Comment by mrserb** [comment]
The testing is in progress.

**Comment by mrserb** [comment]
The new test fails on the headless macs(I cannot reproduce it on my laptop):

----------System.out:(1/17)----------
ErrorCount:53,53
----------System.err:(13/1055)----------
java.lang.RuntimeException: Incorrect mirrored character with java.awt.Font[family=AppleMyungjo,name=AppleMyungjo,style=plain,size=12]
	at MirrorTest.main(MirrorTest.java:147)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
	at java.base/java.lang.Thread.run(Thread.java:831)

Do we need to mark it as headful, or it might be a product bug?

**Comment by toshiona** [comment]
Apologize for the trouble. That's my testcase issue.
That error may occur on macOS High Sierra. Newer OSes (Mojave, Catalina, and Big Sur) may not have the problem. Differences between antialias on and off are bigger than the real error which I'd like to detect.
So, I revised the test to compare the center of gravity of each glyph.
Could you kindly re-review and retest this?

**Review by mrserb — APPROVED** [review]
An updated test works fine.

**Comment by victordyakov** [comment]
@toshiona you can push it as you have 2 approvals. As a reminder today is a last day for fixing P4s in jdk 16

**Comment by toshiona** [comment]
@mrserb @prrace 
Thank you so much for the review. Could you sponsor this patch?

**Pull request closed without merging**